### PR TITLE
bpo-30006 More robust concurrent.futures.ProcessPoolExecutor

### PR DESCRIPTION
**Pull request by tomMoral**
This set of patches intends to fix some silent deadlocks that can happened with
the present `ProcessPoolExecutor` implementation. The deadlock situations
notably include:

- Pickling and Unpickling errors in task definitions and results.
- Processes that get killed in non pythonic ways (kill -9 / segfault).

The commit message of each patch gives more details but in essence:

  1. Support for multiprocessing context instead of forcing the use of `fork` on
posix systems.
  2. Avoid deadlock when a process dies with the `result_queue.wlock` acquired.
  3. Deadlock free clean up of failures in the `queue_manager_thread` and the
`queue_feeder_thread`.
  4. Deadlock free termination of workers and threads when the
`ProcessPoolExecutor` instance has beeen gc'ed.

The commits are incremental, each one adds new fixes on top of the previous
ones to handle extra deadlock situations. If you think that some changes in the
last commits need more discussion or refinement, let me now so I can separate
them and open different tickets/ PR.

Each commit passes the test suit on its own but the addition of the 3 context
testing makes `test_concurrent_futures` longer (~140s on my computer). One
fix would be to split the test for each `context` in seprate files (allowing test
parallelization). I did not do it as it changes the structure of the test suit but
I can implement it if it is necessary.

This work was done as part of the [`loky`][1] project in collaboration with
@ogrisel. It provides a backport of the same features for older versions of
python including 2.7 for legacy users and reusable `executors`.

[1]: https://github.com/tomMoral/loky

https://bugs.python.org/issue30006

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://cpython-devguide.readthedocs.io/pullrequest.html#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

**Comment by ogrisel** [comment]
I have a hard time understanding the codecov report on this PR but it seems that the way tests are currently run on travis, it is not possible to collect coverage data in part of the code that is only executed in multiprocessing worker processes.

For loky we had to use advanced coverage configuration to collect such data, see for instance:

- https://github.com/tomMoral/loky/blob/master/.coveragerc
- https://github.com/tomMoral/loky/blob/master/tox.ini#L19-L30
- https://github.com/tomMoral/loky/blob/master/continuous_integration/install_coverage_subprocess_pth.py

Please let us know if you want us to adapt such configuration to better collect coverage data on children process code in the cpython project.

**Comment by tomMoral** [comment]
@pitrou Rebasing this PR on top of the current master broke our deadlock detection test on unpicklable tasks.
See [bpo-30414](http://bugs.python.org/issue30414?@ok_message=msg%20295022%20created%0Aissue%2030414%20nosy_count%2C%20nosy%2C%20messages%2C%20message_count%20edited%20ok&@template=item#msg295022)

**Comment by tomMoral** [comment]
~I spotted a design mistake and I forgot to add test for the `Queue`.
I am on it right now.~

File: Lib/test/_test_multiprocessing.py [hunk 1]
```diff
@@ -766,6 +766,33 @@ def __reduce__(self):
...
 
+    def test_queue_feeder_on_queue_feeder_error(self):
+        # bpo-30006: verify feeder handles exceptions using the
+        # _on_queue_feeder_error hook.
+        if self.TYPE != 'processes':
+            self.skipTest('test not appropriate for {}'.format(self.TYPE))
+
+        class NotSerializable(object):
+            def __init__(self):
+                self.pass_test = True
```
Comment (ogrisel) [inline Lib/test/_test_multiprocessing.py]: This should be set to `False` by default, otherwise the test is too easy.

Comment (tomMoral) [reply Lib/test/_test_multiprocessing.py]: If I set it to `False`, the test will always fail as I use `&=` afterward.

Comment (ogrisel) [reply Lib/test/_test_multiprocessing.py]: Ok then to make the test more readable we should have two flags:

- one named `reduce_was_called` set to `False` in `__init__` and subsequently set to True in `__reduce__`
- another named `on_queue_feeder_error_was_called` set to `False` in `__init__` and subsequently set to True in `_on_queue_feeder_error` by the Queue itself if `e` and `obj` have the expected type.

And the reverse the order of  ` q.put(unserializable_obj)` and `q.put(True)` to as to make sure that we can make `self.assertTrue` on the flags after `self.assertTrue(q.get(timeout=0.1))` has succeeded.

Comment (tomMoral) [reply Lib/test/_test_multiprocessing.py]: Ok I did it and also updated the timeout to match the new timeout from #2148

**Comment by tomMoral** [comment]
I updated this PR in the last few days:
- I adapted the code base to deal with the unfailing `Queue` (see [bpo30414](http://bugs.python.org/issue30414?@ok_message=msg%20295022%20created%0Aissue%2030414%20nosy_count%2C%20nosy%2C%20messages%2C%20message_count%20edited%20ok&@template=item#msg295022))
- I removed the update in coverage to avoid adding unrelated changes.

**Comment by tomMoral** [comment]
This PR still requires some small fixes (particularly to avoid multiple test run in travis gcc tests).
But feel free to have a look at the diff if you want to start a discussion.

**Comment by pitrou** [comment]
@tomMoral, do you think it would be possible to split this PR into several ones based on the different issues being fixed (or improvements being made)? I think I'll have a hard time reviewing the PR as is, given how delicate multiprocessing (or concurrent.futures) code generally is.

**Comment by tomMoral** [comment]
@pitrou Yes no problem. I created a PR #3682 for the first part which allow to pass a context in the `ProcessPoolExecutor` constructor. I will create other PR for each part when I get a bit more time but we can start with this one.

**Comment by tomMoral** [comment]
@pitrou I created a second PR #3895 for the third commit (I realised the second commit was not necessary). It should fix the interpreter freeze caused by pickling/unpickling errors.

**Comment by ogrisel** [comment]
I think we can close this PR in favor of #3895 and #4256.

**Comment by pitrou** [comment]
Ok, closing.

**Pull request closed without merging**